File: quicksort.py

def quicksort(arr, start, end):
    
    if (start<end):
        # print(arr)
        index = partition(arr,start, end)
        # print('\n' + "left ")
        quicksort(arr,start, index-1)
        # print('\n' + "right ")
        quicksort(arr, index+1, end)
        

def swap(arr,a,b):
    temp = arr[a]
    arr[a] = arr[b]
    arr[b] = temp

def partition(arr,start, end):
    # print("Partition ")
    # print(arr[start:end+1])
    
    pivot = arr[end]
    i = start
    for j in range (i,end):
        if (arr[j] < pivot):
            swap(arr, j, i)
            i+=1
    
    swap(arr, i, end)
    # print("``````````")
    # print(arr[start:end+1])
    # print ('\n' + str(i) + '\n')

    return i


if __name__ =='__main__':
    arr = [51,4,6,2,1,21]
    
    quicksort(arr,0,len(arr)-1)
   
    print("main ")
    print(arr)

File: quicksort.cpp
#include<iostream>
#include<bits/stdc++.h>

using namespace std;

void swap(int *a, int *b)
{
    int temp = *a;
    *a = *b;
    *b = temp;
}

int Partition(int arr[], int start, int end)
{
    int pivot = arr[end];
    int i = start;
    
    for(int j = i; j <= end-1 ; j++)
    {
        if (arr[j] < pivot)
        {
            swap(&arr[i], &arr[j]);
            i++;
        }
    }

    swap(&arr[i], &arr[end]);
    return i;
}

void quicksort(int arr[], int start, int end)
{
    if (start<end)
    {
        int index = Partition(arr, start, end);
        quicksort(arr,start, index-1);
        quicksort(arr, index+1, end);
    }
}

int main()
{
    int arr[6] = {5,4,3,2,1,0};
   
    quicksort(arr,0, 5);

    for(int i = 0; i < 6 ; i++)
    {
        cout << arr[i] << " ";
    }
}